test/test_wiki_trac_rst_convert.py

```diff
@@ -0,0 +1,105 @@
+import unittest
+
+from wiki_trac_rst_convert import convert_content
+
+
+class TracRstToVanillaRst(unittest.TestCase):
+    """
+    Test conversion of content from Trac-flavored reStructuredText to
+    vanilla reStructuredText, that is supported by GitHub
+    """
+
+    def assertConvertedContent(self, expected: str, source: str):
+        """
+        Run the Trac RST `source` through the converter,
+        and assert that the output equals `expected`.
+        """
+        self.assertEqual(expected, convert_content(source))
+
+    def test_empty(self):
+        """
+        An empty string is appended a blank line.
+
+        A file not ending in a newline character is not POSIX compliant,
+        and may result in complaints from programs, like `git diff`
+        saying "\ No newline at end of file".
+        https://stackoverflow.com/a/729795/235463
+        """
+        self.assertConvertedContent('\n', '')
+
+    def test_newline(self):
+        """
+        A newline should not be appended another newline.
+        """
+        self.assertConvertedContent('\n', '\n')
+
+    def test_removes_rst_wrapping(self):
+        """
+        The Trac wiki syntax requires reStructuredText to be wrapped in
+        {{{ #!rst }}} markers.
+        They must be removed from the output.
+        """
+        self.assertConvertedContent('\n', '{{{#!rst}}}')
```

Comment:
It's usually a bad practice to have multiple use cases in a single test, but maybe also add variants like
```suggestion
        self.assertConvertedContent('\n', '{{{#!rst}}}')
        self.assertConvertedContent(
            '\n',
            '\n'
            '{{{\n'
            '#!rst'
            '\n'
            '}}}')
```

And even some other combinations.

Reply:
I know it's a bad practice typically, but I agree that it would be easier to maintain like this; each test is so tiny.

Still, we need to make sure it's clear *which* example failed, if one did fail.

Also, none of my tests have the "Arrange - Act - Assert" structure as per [Structure of a test](https://geck.chevah.com/programming/testing.html#structure-of-a-test). This avoids a lot of duplication (I agreed with your suggestion of creating `assertConvertedContent`); and it is still understandable, because the `convert_content` function is pure.

I suspect "Arrange - Act - Assert" is much more relevant when dealing with stateful code (with side-effects).

Reply:
What you have here are `parametrized` tests

https://docs.pytest.org/en/stable/parametrize.html

but I prefer to not use the decorator... and just have docstrings and maybe inline comments for each example.

in terms of knowing which test fails... it's true.
But here we have unit tests.... and we can write the assertions from very basic first to very complex last.